### PR TITLE
Improve error handling rds.py for RDS cluster and snapshot

### PR DESCRIPTION
**Pull request by alinabuzachis**
##### SUMMARY

 - Improve error handling rds.py for `rds_cluster`
 - Add handlers for `rds_instance_snapshot` and `rds_cluster_snapshot`

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

module_utils/rds.py

### CODE REVIEW

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

File: plugins/module_utils/rds.py [hunk 1]
```diff
@@ -181,20 +230,43 @@ def wait_for_cluster_status(client, module, db_cluster_id, waiter_name):
         module.fail_json_aws(e, msg="Failed with an unexpected error while waiting for the DB cluster {0}".format(db_cluster_id))
 
 
+def wait_for_snapshot_status(client, module, db_snapshot_id, waiter_name):
```
Comment (jillr) [inline plugins/module_utils/rds.py:233]: Could you please create a new unit test file for this module_util and cover this new function?  Eventually we should try to cover the rest of the util but it's a lot easier to ask for folks to contribute tests for changes if the test file already exists.

We should also run the integration tests for the RDS modules with a depends-on for this change, to ensure there's no regressions.

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by goneri** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
cc @goneri 

```
2021-12-20 10:36:56.079435 \| fedora-35 \| TASK [aws_s3 : make tempfile 4 GB for linux] ***********************************
--
5391 | 2021-12-20 10:36:56.079637 \| fedora-35 \| task path: /home/zuul/.ansible/collections/ansible_collections/amazon/aws/tests/integration/targets/aws_s3/tasks/main.yml:499
5392 | 2021-12-20 10:36:56.244626 \| fedora-35 \| Using module file /tmp/ansible-test-qyqzaib3/ansible/modules/command.py
5393 | 2021-12-20 10:36:56.244945 \| fedora-35 \| Pipelining is enabled.
5394 | 2021-12-20 10:36:56.245217 \| fedora-35 \| <testhost> ESTABLISH LOCAL CONNECTION FOR USER: zuul
5395 | 2021-12-20 10:36:56.245359 \| fedora-35 \| <testhost> EXEC /bin/sh -c 'ANSIBLE_DEBUG_BOTOCORE_LOGS=True /home/zuul/venv/bin/python && sleep 0'
5396 | 2021-12-20 10:36:57.054748 \| fedora-35 \| fatal: [testhost]: FAILED! => {
5397 | 2021-12-20 10:36:57.054840 \| fedora-35 \|     "changed": true,
5398 | 2021-12-20 10:36:57.054850 \| fedora-35 \|     "cmd": [
5399 | 2021-12-20 10:36:57.054856 \| fedora-35 \|         "dd",
5400 | 2021-12-20 10:36:57.054862 \| fedora-35 \|         "if=/dev/zero",
5401 | 2021-12-20 10:36:57.054868 \| fedora-35 \|         "of=/tmp/ansible.6d71q1je/largefile",
5402 | 2021-12-20 10:36:57.054875 \| fedora-35 \|         "bs=1M",
5403 | 2021-12-20 10:36:57.054880 \| fedora-35 \|         "count=4096"
5404 | 2021-12-20 10:36:57.054886 \| fedora-35 \|     ],
5405 | 2021-12-20 10:36:57.054892 \| fedora-35 \|     "delta": "0:00:00.601860",
5406 | 2021-12-20 10:36:57.054897 \| fedora-35 \|     "end": "2021-12-20 10:36:57.021668",
5407 | 2021-12-20 10:36:57.054903 \| fedora-35 \|     "invocation": {
5408 | 2021-12-20 10:36:57.054909 \| fedora-35 \|         "module_args": {
5409 | 2021-12-20 10:36:57.054915 \| fedora-35 \|             "_raw_params": "dd if=/dev/zero of=/tmp/ansible.6d71q1je/largefile bs=1M count=4096",
5410 | 2021-12-20 10:36:57.054928 \| fedora-35 \|             "_uses_shell": false,
5411 | 2021-12-20 10:36:57.054935 \| fedora-35 \|             "argv": null,
5412 | 2021-12-20 10:36:57.054941 \| fedora-35 \|             "chdir": null,
5413 | 2021-12-20 10:36:57.054947 \| fedora-35 \|             "creates": null,
5414 | 2021-12-20 10:36:57.054953 \| fedora-35 \|             "executable": null,
5415 | 2021-12-20 10:36:57.054960 \| fedora-35 \|             "removes": null,
5416 | 2021-12-20 10:36:57.054966 \| fedora-35 \|             "stdin": null,
5417 | 2021-12-20 10:36:57.054972 \| fedora-35 \|             "stdin_add_newline": true,
5418 | 2021-12-20 10:36:57.054978 \| fedora-35 \|             "strip_empty_ends": true,
5419 | 2021-12-20 10:36:57.054984 \| fedora-35 \|             "warn": false
5420 | 2021-12-20 10:36:57.055001 \| fedora-35 \|         }
5421 | 2021-12-20 10:36:57.055007 \| fedora-35 \|     },
5422 | 2021-12-20 10:36:57.055012 \| fedora-35 \|     "msg": "non-zero return code",
5423 | 2021-12-20 10:36:57.055017 \| fedora-35 \|     "rc": 1,
5424 | 2021-12-20 10:36:57.055023 \| fedora-35 \|     "start": "2021-12-20 10:36:56.419808",
5425 | 2021-12-20 10:36:57.055030 \| fedora-35 \|     "stderr": "dd: error writing '/tmp/ansible.6d71q1je/largefile': No space left on device\n947+0 records in\n946+0 records out\n992100352 bytes (992 MB, 946 MiB) copied, 0.596905 s, 1.7 GB/s",
5426 | 2021-12-20 10:36:57.055037 \| fedora-35 \|     "stderr_lines": [
5427 | 2021-12-20 10:36:57.055043 \| fedora-35 \|         "dd: error writing '/tmp/ansible.6d71q1je/largefile': No space left on device",
5428 | 2021-12-20 10:36:57.055049 \| fedora-35 \|         "947+0 records in",
5429 | 2021-12-20 10:36:57.055054 \| fedora-35 \|         "946+0 records out",
5430 | 2021-12-20 10:36:57.055060 \| fedora-35 \|         "992100352 bytes (992 MB, 946 MiB) copied, 0.596905 s, 1.7 GB/s"
5431 | 2021-12-20 10:36:57.055066 \| fedora-35 \|     ],
5432 | 2021-12-20 10:36:57.055071 \| fedora-35 \|     "stdout": "",
5433 | 2021-12-20 10:36:57.055077 \| fedora-35 \|     "stdout_lines": []
5434 | 2021-12-20 10:36:57.055083 \| fedora-35 \| }
```

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

File: plugins/modules/ec2_vpc_igw.py [hunk 2]
```diff
@@ -217,7 +217,8 @@ def ensure_igw_present(self, vpc_id, tags, purge_tags):
             self._connection, self._module, igw['internet_gateway_id'],
             resource_type='internet-gateway', tags=tags, purge_tags=purge_tags
         )
-        igw_info = self.get_igw_info(self.get_matching_igw(vpc_id), vpc_id)
+        igw = self.get_matching_igw(vpc_id)
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vpc_igw.py:220]: This fixes “TypeError: 'NoneType' object is not subscriptable”.

File: plugins/module_utils/rds.py (hunk 1 above)
Comment (alinabuzachis) [reply plugins/module_utils/rds.py:233]: @jillr Done that!

**Review by markuman — APPROVED** [review]
LGTM.

Any idea for what the 5th waiter is good for? https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds.html#RDS.Waiter.DBSnapshotCompleted

**Comment by alinabuzachis** [comment]
> LGTM.
> 
> Any idea for what the 5th waiter is good for? https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds.html#RDS.Waiter.DBSnapshotCompleted

I guess it's related to the snapshot progress. Not sure about the available statuses for db_instance_snapshot, but I found those for db_cluster_snapshot https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds.html#RDS.Client.create_db_cluster_snapshot

```
Status (string) --

Specifies the status of this DB cluster snapshot. Valid statuses are the following:

available
copying
creating
```

File: plugins/modules/ec2_vpc_igw.py (hunk 2 above)
Comment (jatorcasso) [reply plugins/modules/ec2_vpc_igw.py:220]:
```suggestion

```
@alinabuzachis pretty sure this line can be removed as `igw` is defined earlier on in the method.. worked for me at least!

**Review by jillr — APPROVED** [review]

**Comment by alinabuzachis** [comment]
regate

**Comment by alinabuzachis** [comment]
regate

**Comment by alinabuzachis** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Comment by abikouo** [comment]
recheck

**Comment by goneri** [comment]
recheck

**Comment by alinabuzachis** [comment]
recheck

**Pull request merged**